### PR TITLE
meson: use wraps and fallbacks for some dependencies

### PR DESCRIPTION
**Pull request by andy5995**
For #95

Also, if lua is installed on Linux, it will just link dynamically instead of building.

### CODE REVIEW

**Comment by andy5995** [comment]
Here's what happens when lua isn't found: https://github.com/netpanzer/netpanzer/actions/runs/7548411495/job/20550403662#step:8:32

**Comment by andy5995** [comment]
The resulting binary always links dynamically to the dependencies. I'm not sure how to handle this so I've opened up a [discussion on the meson repo](https://github.com/mesonbuild/meson/discussions/12754)

File: meson.build
Comment (eli-schwartz) [inline meson.build:57]: There's no reason to look this up twice, especially, the options specified the second time won't be used as the first lookup has to succeed.

File: meson.build
Comment (eli-schwartz) [inline meson.build:74]: ... and here.

File: meson.build
Comment (eli-schwartz) [inline meson.build:91]: ... and here.

File: meson.build
Comment (eli-schwartz) [inline meson.build:109]: ... and here.

File: meson.build
Comment (eli-schwartz) [inline meson.build:125]: ... and here.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: meson.build
Comment (eli-schwartz) [inline meson.build:117]: I think that you also want to look up sdl2 before the sdl2_* dependencies. Otherwise, sdl2 itself might be internally looked up by SDL2_ttf or SDL2_mixer, and meson will already have a cached dependency which doesn't use your default_options.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

**Comment by eli-schwartz** [comment]
It might also be easier to set `'freetype:default_library=static'` either on the command line or in the main `project(..., default_options: [....])` to ensure the option is handled even when some nested dependency is the thing that actually invokes `dependency()`.

**Comment by andy5995** [comment]
The problem now is I get 2 undefined errors when building the tests. Not sure if I need to make some adjustments to where [the netpanzer lib is created](https://github.com/netpanzer/netpanzer/blob/master/src/NetPanzer/meson.build) or something else. Probably a combination ;)

**Comment by andy5995** [comment]
> The problem now is I get 2 undefined errors when building the tests. Not sure if I need to make some adjustments to where [the netpanzer lib is created](https://github.com/netpanzer/netpanzer/blob/master/src/NetPanzer/meson.build) or something else. Probably a combination ;)

Actually the tests build and run fine in Linux, so it must be something else.

**Comment by andy5995** [comment]
Ready for review and testing on Windows. @winrid if you want to wait until after the rc7 release to review and merge this, I don't mind.

The 2 test programs fail to build. I think it might be related to https://github.com/mesonbuild/meson/issues/10472 but really I don't understand enough to say it's a reasonable assumption. Bottom line is I don't think we need to hold this up just because the two test programs won't build.

**Comment by andy5995** [comment]
When I build in a container on Linux, with static linking, then try to run it on the host, I get:

```
[andy@prometheus docker3_build](meson-add-lua)$ ./netpanzer
2024-01-19 00:51:32 dataDir: /home/andy/src/netpanzer/netpanzer

2024-01-19 00:51:32 ERROR in '@/home/andy/.netpanzer//config/client.cfg':48: config 'anticheat' is invalid.
2024-01-19 00:51:32 Couldn't initialize sound: SDL_Init audio error: No available audio device
2024-01-19 00:51:32 Initialisation failed:
Couldn't initialize SDL_video subsystem: No available video device
2024-01-19 00:51:32 Couldn't initialize the game: bootstrap failed.
```

**Comment by winrid** [comment]
I think wraps are a nice abstraction, just have to test IDE integration (can I still easily go to source of method etc). LMK when ready to test

**Comment by eli-schwartz** [comment]
@andy5995 you may need to set other sdl2 configuration options based on what devices you want it to support.

@winrid this should work fine assuming your ISE uses for example clang's `compile_commands.json` database. The definitions from the subproject should be in the include directory arguments for each file.

**Comment by andy5995** [comment]
> @andy5995 you may need to set other sdl2 configuration options based on what devices you want it to support.

K, I found that in https://mesonbuild.com/Build-options.html#using-build-options

@eli-schwartz is there a way to set the sdl2 defaults in the superproject meson.build?

> I think wraps are a nice abstraction, just have to test IDE integration (can I still easily go to source of method etc). LMK when ready to test

@winrid Yes, ready to test.

File: src/Lib/2D/Surface.hpp
Comment (winrid) [inline src/Lib/2D/Surface.hpp:21]: This is worse - we should be namespacing our dependencies on include.

**Comment by winrid** [comment]
Confirmed IDE integration works. However, still do not like non-name spaced includes. :)

File: src/Lib/2D/Surface.hpp
Comment (andy5995) [reply src/Lib/2D/Surface.hpp:21]: I'm not sure how to handle that. On Slackware, Ubuntu, Manjaro, Debian, `sdl2-config --cflags` returns `# sdl2-config --cflags
-I/usr/include/SDL2 -D_REENTRANT` so the compiler is going to look in SDL2/, therefore, `#include <SDL.h>` is all that's needed in that case. Build systems typically call `sdl2-config` to get the right flags for programs requiring sdl2.

When the source is download in the case of fallbacks with meson, the header is in 'subprojects/SDL2-2.28.5/include'

Comment (winrid) [reply src/Lib/2D/Surface.hpp:21]: Okay! I guess it's fine for now until it's an issue, don't want to cause extra work if it's a lot of trouble.

**Comment by andy5995** [comment]
@eli-schwartz Now something is acting like a bug. I used `-Dsdl2:use_audio_alsa=enabled` and got an error that alsa wasn't found. I tried disabling and ran `meson setup --wipe` (In the docker container that's been working fine up until tried changing the also option);

```
sdl2| Project name: sdl2
sdl2| Project version: 2.28.5
sdl2| C compiler for the host machine: cc (gcc 11.4.0 "cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
sdl2| C linker for the host machine: cc ld.bfd 2.38
sdl2| Compiler for C supports arguments -msse3: YES
sdl2| Compiler for C supports arguments -Wno-deprecated-declarations: YES
sdl2| Compiler for C supports arguments -Wdeclaration-after-statement: YES
sdl2| Compiler for C supports arguments -Wshadow: YES (cached)
sdl2| Compiler for C supports arguments -fno-strict-aliasing: YES
sdl2| Compiler for C supports arguments -mpreferred-stack-boundary=2: NO
sdl2| Run-time dependency gl found: NO (tried pkgconfig and system)
sdl2| Run-time dependency opengl found: NO (tried pkgconfig and cmake)
sdl2| Run-time dependency glesv2 found: NO (tried pkgconfig and cmake)
sdl2| Message: Using embedded GLESv2 headers
sdl2| Run-time dependency egl found: NO (tried pkgconfig and cmake)
sdl2| Run-time dependency egl found: NO (tried pkgconfig and cmake)
sdl2| Message: Using embedded EGL headers
sdl2| Run-time dependency alsa found: NO (tried pkgconfig and cmake)

../subprojects/SDL2-2.28.5/meson.build:215:0: ERROR: Dependency "alsa" not found, tried pkgconfig and cmake

A full log can be found at /netpanzer/docker3_build/meson-logs/meson-log.txt
npbuilder@852b1fb05660:/netpanzer/docker3_build$ meson configure | grep audio
npbuilder@852b1fb05660:/netpanzer/docker3_build$ meson configure -Dsdl2:use_audio_alsa=disabled

ERROR: No such build data file as "'/netpanzer/docker3_build/meson-private/build.dat'".
```

On the host when I enter that same directory, `--wipe` completes

File: src/Lib/2D/Surface.hpp
Comment (andy5995) [reply src/Lib/2D/Surface.hpp:21]: my thoughts too. If I knew a better way, I'd either do it, or I'd take a break and do it. ;) But definitely we'll have to change that if it becomes an issue. Note it's passing all CI tests (the Cirrus failure is unrelated).

File: setup-build.bat
Comment (andy5995) [inline setup-build.bat:1]: @eli-schwartz nvm, I see now that sdl2main is available through the sdl2 wrap and implemented it correctly (afaik) in https://github.com/netpanzer/netpanzer/pull/133/commits/fbb6acb85d7140b892790183261a3e2e888b1187#diff-30d8f6be6320feeacf686be94f48c70869b52630e01ea625f0f15adc0d57c3e4R92

**Comment by andy5995** [comment]
@ StanleySweet I remember you had trouble with the boost libraries when you tried using @mesonbuild for @0ad. Had you used meson's wrap system?

File: src/Lib/2D/Surface.hpp
Comment (eli-schwartz) [reply src/Lib/2D/Surface.hpp:21]: > This is worse - we should be namespacing our dependencies on include.

This is an illegal maneuver. The `#include` name of a dependency is exactly no more and no less than whatever the dependency tells you to use.

There is no such concept as namespacing. The "namespacing" here is that the sdl project has moral rights to the name SDL.h, and specified that if your application uses version 2 then you add the SDL2/ directory to your include paths and use `#include <SDL.h>`, and if you use version 3 then you add the root includedir to your include paths and use `#include <SDL3/SDL.h>`.

Yes, the style changed between versions. But even if you have both sdl 2 and 3 brought into your project at the same time (???) version 3's headers won't be available as `<SDL.h>` so there is no clash.

**Comment by StanleySweet** [comment]
Don't remember. Patch is here https://code.wildfiregames.com/D2743 if you wanna have a look.

**Comment by winrid** [comment]
@andy5995 this is currently not working for me on win10:

```
 meson setup _build --wipe
The Meson build system
Version: 1.3.1
Source dir: C:\Users\winrid\Documents\netpanzer      
Build dir: C:\Users\winrid\Documents\netpanzer\_build
Build type: native build
Project name: netpanzer  
Project version: 0.9.0.07
C compiler for the host machine: ccache gcc (gcc 13.1.0 "gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders) 13.1.0")
C linker for the host machine: gcc ld.bfd 2.40
C++ compiler for the host machine: ccache c++ (gcc 13.1.0 "c++ (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders) 13.1.0")
C++ linker for the host machine: c++ ld.bfd 2.40
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C supports arguments -fno-common: YES 
Compiler for C supports arguments -Wshadow: YES 
Compiler for C++ supports arguments -fno-common: YES 
Compiler for C++ supports arguments -Wshadow: YES 
Configuring config.h using configuration
Library ws2_32 found: YES
Did not find pkg-config by name 'pkg-config'
Found pkg-config: NO
Found CMake: C:\MinGW\bin\cmake.EXE (3.26.4)
Run-time dependency lua-5.1 found: NO (tried pkgconfig and cmake)
Has header "lua.h" : NO 
Run-time dependency lua-5.1 found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency lua-5.1

meson.build:52:14: ERROR: Subproject exists but has no meson.build file.

```

**Comment by andy5995** [comment]
@winrid Try removing the srcroot/subprojects/lua-5.1.5 directory and then make sure you call meson setup from from the batch file. e.g. `setup-build.bat <builddir>`

**Comment by winrid** [comment]
ah right, thanks, forgot to use the new batch file.

**Pull request merged**
